[QuotaRP] Split QuotaLimits API to - Usages and Quotas API

rahuls-microsoft opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [x] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [x ] Other, please clarify - Adding Network RP examples
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ x] SDK of .NET (need service team to ensure code readiness)
      - [ x] SDK of Python
      - [ x] SDK of Java
      - [x ] SDK of Js
      - [x ] SDK of Go
      - [x ] PowerShell
      - [x ] CLI
      - [ x] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

rahuls-microsoft committed:
1.Breaking Change - Removing duplicate autoQuotaIncreaseState from supportTicketAction.
2. Removing state=enabledForPreview.
3. Updating Example.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Updating Provider name in example.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into master

rahuls-microsoft committed:
Capacity RP Quota APIs for GA

rahuls-microsoft committed:
Adding preview version to stable dir as base files.

rahuls-microsoft committed:
Files from preview version, which will be removed.

rahuls-microsoft committed:
Updating to GA Version.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
1. Updated reservation version tag to - Microsoft.Capacity/preview/2020-10-01-preview/reservations.json, which matches the previous SDK release.
2. Documentation changes.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Documentation updates for put and patch request, which includes only ONE quota request, so changing quotas to quota.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

rahuls-microsoft committed:
Update documenation for put.

rahuls-microsoft committed:
Doc update.

rahuls-microsoft committed:
Adding -preview tag for GO sdk.

rahuls-microsoft committed:
Updating Go settings for output folder to be -preview.

rahuls-microsoft committed:
Adding preview subdir for GO package.

rahuls-microsoft committed:
Adding preview subdir to Go sdk path.

rahuls-microsoft committed:
Removing tag - tag: package-2020-10-25-preview

ArcturusZhang committed:
introduce a new tag to keep the go sdk package unchanged

ArcturusZhang committed:
Revert "introduce a new tag to keep the go sdk package unchanged"

This reverts commit 8f58d3451e9ae5a4746d2a5883911f14ca261f1b.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Adding Azure CLI readme configuration.

rahuls-microsoft committed:
Updating the ReadMe configurations.

rahuls-microsoft committed:
Updating readme.md

rahuls-microsoft committed:
Removing duplicate cli: tag.

rahuls-microsoft committed:
Fixing build issues.

rahuls-microsoft committed:
Removing Parameter alias.

rahuls-microsoft committed:
cli config changes.

rahuls-microsoft committed:
Updating the package name to reservation.

rahuls-microsoft committed:
Updates - removin .az.md.

rahuls-microsoft committed:
Updating Example names to be unique for Get PUT

rahuls-microsoft committed:
Adding operations and python-sdk-output-folder.

rahuls-microsoft committed:
Updates.

rahuls-microsoft committed:
Removing unused file.

rahuls-microsoft committed:
Revert "Removing unused file."

This reverts commit 4f763bdb14240a3f1ba8e1dce6c7f0b643ee6a9e.

rahuls-microsoft committed:
Revert "Revert "Removing unused file.""

This reverts commit 60c520e5b50bc13c42d8227624f32031c7ca7694.

rahuls-microsoft committed:
Add reference to python readme.md file.

rahuls-microsoft committed:
Merge branch 'master' into cli

rahuls-microsoft committed:
Adding python readme file.

rahuls-microsoft committed:
Adding Operations for test-scenario

rahuls-microsoft committed:
Adding split

rahuls-microsoft committed:
Disabling few operation, which is not needed.

rahuls-microsoft committed:
Removing split for Operation

rahuls-microsoft committed:
Disabling the reservation Operations.

rahuls-microsoft committed:
Disable reservation operations.

rahuls-microsoft committed:
Revert "Disable reservation operations."

This reverts commit d98cbcbafeb855395f9de45292dfc9db2f5aea9d.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Adding client-subscription-bound: false

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

rahuls-microsoft committed:
Adding Network RP examples and references.

rahuls-microsoft committed:
Updating Network Put and Patch requests.

rahuls-microsoft committed:
Updating GetQuotaRequests example.

rahuls-microsoft committed:
Fixing formatting.

rahuls-microsoft committed:
Updating examples.

rahuls-microsoft committed:
Adding error details for the quota request status.

rahuls-microsoft committed:
Updating examples

rahuls-microsoft committed:
Fixing formatting

rahuls-microsoft committed:
Fixing model definition.

rahuls-microsoft commented:
> NewApiVersionRequired reason:
> 
> A service’s API is a contract with customers and is represented by using the api-version query parameter. Changes such as adding an optional property to a request/response or introducing a new operation is a change to the service’s contract and therefore requires a new api-version value. This is critically important for documentation, client libraries, and customer support.
> 
> EXAMPLE: if a customer calls a service in the public cloud using api-version=2020-07-27, the new property or operation may exist but if they call the service in a government cloud, air-gapped cloud, or Azure Stack Hub cloud using the same api-version, the property or operation may not exist. Because there is no clear relationship between the service api-version and the new property/operation, customers can’t trust the documentation and Azure customer have difficulty helping customers diagnose issues. In addition, each client library version documents the service version it supports. When an optional property or new operation is added to a service and its Swagger, new client libraries must be produced to expose this functionality to customers. Without updating the api-version, it is unclear to customers which version of a client library supports these new features.

The API is still in private preview for Test In Production. No customer or users have been onboarded.

rahuls-microsoft commented:
The API is still in private preview for Test In Production. No customer or users have been onboarded.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

raych1 commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/examples/putMachineLearningServicesQuotaRequestLowPriority.json:
```diff
@@ -29,9 +33,7 @@
-        "location": "https://management.azure.com/subscriptions/D7EC67B3-7657-4966-BFFC-41EFD36BAAB3/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/quotaLimitsRequests/2B5C8515-37D8-4B6A-879B-CD641A2CF605?api-version=2021-03-15-preview",
-        "Retry-After": "30",
-        "Azure-AsyncOperation": "https://management.azure.com/subscriptions/D7EC67B3-7657-4966-BFFC-41EFD36BAAB3/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/operationsStatus/2B5C8515-37D8-4B6A-879B-CD641A2CF605?api-version=2021-03-15-preview"
+        "location": "https://management.azure.com/subscriptions/D7EC67B3-7657-4966-BFFC-41EFD36BAAB3/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/quotaLimitsRequests/2B5C8515-37D8-4B6A-879B-CD641A2CF605?api-version=2021-03-15-preview"
```
Why this one return `location` header instead of `azure-asyncoperation`?

rahuls-microsoft replied:
Based on - https://github.com/Azure/azure-sdk-for-net/issues/3020
if Azure-AsyncOperation was polled and the operation succeeded then the Location header should be retrieved (if possible) in order to get the correct response code.
For 202 Azure-AsyncOperation is checked.
When the operation is completed, then 200 is returned with Location header.

rahuls-microsoft committed:
Revert "Fixing model definition."

This reverts commit cef9f52f142cf6c420a5f89bd0e2d806e29530c1.

rahuls-microsoft committed:
Updating example.

rahuls-microsoft committed:
Making ResoureType as String, as it will change with different RPs and all possible values cannot be added in advance.

rahuls-microsoft commented:
@raych1 Please review the changes.
The API is still in Private Preview for our Testing.
Minor changes -
1. Adding Network examples.
2. Making resourceType as string, so that each RPs resourceType can be included with swagger changes.

Please let me know, if you have questions.

rahuls-microsoft committed:
Adding Location header to Put and Path calls 200 response.

rahuls-microsoft committed:
Adding example for Network StandardSkuPublicIpAddresses

rahuls-microsoft committed:
Fixing spelling

rahuls-microsoft committed:
Adding error code to Failed requests and updating examples.

rahuls-microsoft committed:
Updating error message and code.

rahuls-microsoft commented:
We are adding Error code and description for better user experience.
"error": {
-          "description": "Error details of the quota request.",
-          "$ref": "#/definitions/ServiceErrorDetail"
         }
There are no Users onboarded, so updating the respose.
Please let me know, if you have any questions regarding the BreakingChange.
Thanks.

rahuls-microsoft committed:
Updating example for Network - StandardSkuPublicIpAddresses

raych1 commented:
> @raych1 - Please let me know, if there is any open issues?
> How do I download the GO and .net SDKs? We would like to verify the generated SDKs.

@rahuls-microsoft , please go thru breaking change review process if you want to add error code in same version of swagger. BTW, what document do you reference for adding `location` header in 200 response.

rahuls-microsoft commented:
> > @raych1 - Please let me know, if there is any open issues?
> > How do I download the GO and .net SDKs? We would like to verify the generated SDKs.
> 
> @rahuls-microsoft , please go thru breaking change review process if you want to add error code in same version of swagger. BTW, what document do you reference for adding `location` header in 200 response.

1. For adding error object as -
"error" : {
        "code": "{error code}",  
        "message": "{error description}"
    }
https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations#azure-asyncoperation-request-and-response
2. For long running operation Location header -
Based on - https://github.com/Azure/azure-sdk-for-net/issues/3020

if Azure-AsyncOperation was polled and the operation succeeded then the Location header should be retrieved (if possible) in order to get the correct response code.
For 202 Azure-AsyncOperation is checked.
When the operation is completed, then 200 is returned with Location header.

rahuls-microsoft commented:
The schema and API is going to change based on ARM team's suggestions. So, putting in draft mode.

rahuls-microsoft committed:
Adding oneOf construct. If supported, then I will update the examples.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

rahuls-microsoft committed:
Adding discriminator for Limit.

rahuls-microsoft committed:
Adding discriminator to limit.

rahuls-microsoft committed:
Updating network UsagesValue example for discriminator

rahuls-microsoft committed:
Adding limitType definition.

rahuls-microsoft committed:
Adding type to the Limit objects.

rahuls-microsoft committed:
Updating type and example for discriminator

rahuls-microsoft committed:
Updating type for discriminator

rahuls-microsoft committed:
fixinf type definition for discriminator

rahuls-microsoft committed:
Updating schema for discriminator and example.

rahuls-microsoft committed:
Update network example.

rahuls-microsoft committed:
Fixing network example.

rahuls-microsoft committed:
Fixing schema.

rahuls-microsoft committed:
Fixing network example.

rahuls-microsoft committed:
Fix example.

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

rahuls-microsoft committed:
Updating schema.

rahuls-microsoft committed:
Updating schema and example.

rahuls-microsoft committed:
Update example.

rahuls-microsoft committed:
Examples update.

rahuls-microsoft committed:
Update

rahuls-microsoft committed:
Updated schema for discriminator

rahuls-microsoft committed:
Adding value as required.

rahuls-microsoft committed:
Schema Update.

rahuls-microsoft committed:
Replace Limt with LimitObject to fix reference.

rahuls-microsoft committed:
Updates.

rahuls-microsoft committed:
Updates.

rahuls-microsoft committed:
Updating Put/Ptach examples.

rahuls-microsoft committed:
Updating erxamples.

rahuls-microsoft committed:
Schema updated for requestStatus.

rahuls-microsoft committed:
Schema Update.

rahuls-microsoft committed:
Updates to QuotaRequestStatus example.
Change api to quota. Add usages API and update examples reference.

rahuls-microsoft committed:
1. Split quotas and usages apis.
2. Add and update examples for quotas and usages.

rahuls-microsoft committed:
Updating example filename references

rahuls-microsoft committed:
Update Put example and operations id.

rahuls-microsoft committed:
Updating limiType field to limitObjectType

rahuls-microsoft committed:
Updating "limitType" to "limitObjectType"

rahuls-microsoft committed:
Updating Get Examples.
Adding extra property to quotas/limit object.

rahuls-microsoft committed:
1. Updating tags for usages.
2. Rename LimitObjectType to LimitObjectTypes.

rahuls-microsoft committed:
Adding  "limitType": "Independent"

rahuls-microsoft committed:
Updating Operations Tag.

rahuls-microsoft committed:
Updating subscription ids.

rahuls-microsoft committed:
Updating Operations Name.

rahuls-microsoft committed:
Updating the documentations and examples.

rahuls-microsoft committed:
Updates.

rahuls-microsoft committed:
Updates.

rahuls-microsoft commented:
https://msazure.visualstudio.com/One/_workitems/edit/10588364 - Added breaking change justification.

rahuls-microsoft committed:
Updating Quota reference to base object type.

pilor commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/quota.json:
```diff
@@ -342,12 +456,12 @@
       }
     },
-    "/providers/Microsoft.Quota/quotaLimitProviders": {
+    "/providers/Microsoft.Quota/quotaProviders": {
```
This API is not behind a feature flag in your published RP manifest. A name change here is a breaking change as the API is technically public. Please discuss this with the breaking change review folks

rahuls-microsoft replied:
This APIs scope is not limited to user, As it lists what RPs are supported and what dimensions are required for the RP.
I can restore the API name.

rahuls-microsoft replied:
Restored the name to quotaLimitProviders

pilor commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/quota.json:
```diff
@@ -34,16 +34,118 @@
+          "200": {
+            "description": "OK. The usages information will be returned in the response.",
+            "schema": {
+              "$ref": "#/definitions/CurrentQuotaLimitBase"
```
Why does usages return the "currentValue" in a property named "limit"? That isn't user friendly and doesnt seem necessary

rahuls-microsoft replied:
I will change it to "usages" - I will the polymorphism for future extension of this property, when the usages values are combined per SKU family, rather that individual SKU.
"usages": {
            "limitObjectType": "UsagesValue",
            "value": 10
          },

rahuls-microsoft replied:
Separated out quotas and usages.

rahuls-microsoft replied:
> Separated out quotas and usages. usages will be listed under "usages" object, with optional field -
"UsagesTypes": {
      "description": "The quota or usages limit types.",
      "enum": [
        "Individual",
        "Combined"
      ],

pilor commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/examples/getComputeOneSkuUsages.json:
```diff
@@ -7,11 +7,14 @@
       "body": {
-        "id": "subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/quotaLimits/standardNDSFamily",
-        "type": "Microsoft.Quota/quotaLimits",
+        "id": "subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/Quotas/standardNDSFamily",
```
IDs should always have a leading '/'

rahuls-microsoft replied:
I will add leading / to examples.

pilor commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/examples/putComputeOneSkuQuotaRequest.json:
```diff
@@ -15,29 +18,37 @@
     },
     "200": {
+      "headers": {
+        "location": "https://management.azure.com/subscriptions/D7EC67B3-7657-4966-BFFC-41EFD36BAAB3/providers/Microsoft.Compute/locations/eastus/providers/Microsoft.Quota/quotaRequests/2B5C8515-37D8-4B6A-879B-CD641A2CF605?api-version=2021-03-15-preview"
```
I believe I commented on this in a previous PR but a "location" header should keep returning 202 until the async operation is complete. If you point at quotaRequests they will get a 200 back with a quotaRequests resource immediately and think the quota operation is complete.

rahuls-microsoft replied:
I will put the operations url for the location header, which will return 202, until the operation is completed.

pilor replied:
yes, from an async perspective that is fine.  I think you still have a discoverability problem here.  How does a user discover the quotaRequests resource related to the PUT they did on a quotas resource?  There should be some tie between the two.  Maybe as simple as referencing the "quotas" resource ID that a quotaRequests is related to in the quotaRequests response body (and allowing filtering on that property in the collection GET)

rahuls-microsoft replied:
The way it's implemented -
1. User calls the Put API, return is 202 with the Azure-AsyncOperation" : operations url.
Operations API processing:
1. If request is in progress, it returns "Azure-AsyncOperation" with operations api.
2.  User or SDK keeps on checking the operationsStatus API, based on retry-after period. 
3. When the provisioning state is terminal state - failed or succeeded. In the response the "location" header is set to the quotaLimitRequests and  "Azure-AsyncOperation" header is not set.
4. The SDK and ARM template calls the url in location for the quotaLimitRequests and displays the result based on provisioning State. 

I don't see a case, where the PUT or patch api is going to return 200 for long running operation.

pilor replied:
At the end of an async operation the "location" header must point to the result of the request.  Showing a different resource when the user did a PUT on a quotas resource violates that (see # 5 here:  https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#202-accepted-and-location-headers).  202 is fine for PATCH but 200/201 + provisioningState + azure-asyncoperation is the preferred model for async PUTs (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creatingupdating-using-put).  

You CAN put the resource ID of the quotaRequests resource in the "properties" of the azure-asyncoperation operationStatuses resource but I don't know how discoverable that is.  You could try it and see what that experience is like (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#azure-asyncoperation-resource-format)

rahuls-microsoft replied:
The operations status is following the guidelines at - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#azure-asyncoperation-resource-format.

"id": "/subscriptions/id/locations/westus/operationsStatuses/sameguid - is the requestId. Which returns the-
"status" : "RP defined values | Succeeded | Failed | Canceled",
Request details  in properties: {what quota was requested}
In case of error, error details in -
error" : {
    /\* RP must return the *code* and *messages* fields. Please use the schema for the "ErrorResponse" Type from the Common Types definition in the Azure Rest API Specifications [repository](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json) \*/
    "code": "BadArgument",
    "message": "The provided database &#39;foo&#39; has an invalid username."
  }

The ARM template and Azure SDK works fine, so looks like we are following the right guidelines for operationsStatus.
Example from ARM template deployment:
VERBOSE: 10:57:23 AM - Resource Microsoft.Quota/quotaLimits 'PublicIPAddresses' provisioning status is running
And finally it shows Failed (with error details) or Succeeded.

Does it look fine?

pilor replied:
You are combining a few async patterns into one but not following the requirements for any explicitly.  If you are returning a 202 you MUST return a location header (https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creating-or-updating-resources-asynchronously).  You can ALSO return an azure-asyncoperation header if you wish but that is secondary.  ARM's async operation tracking code is more lenient than the spec but the spec must be followed so users that are tracking it using their own code that is explicitly following the spec don't fail

rahuls-microsoft replied:
I have updated the 202 response and headers based on - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creating-or-updating-resources-asynchronously.

Only location header is returned, which points to operationsStatus for 202 and quotaRequests for 200.
When the operation reaches the terminal state, then the location header will point to quotaRequests URL, which will have all the error details or success message.

pilor commented:
@grzegorzzygmunt do Quota APIs fall under the umbrella of APIs you would like to review?

pilor commented on specification/quota/resource-manager/Microsoft.Quota/preview/2021-03-15-preview/quota.json:
```diff
@@ -655,9 +755,113 @@
+        "modelAsString": true
+      }
+    },
+    "LimitJsonObject": {
```
I don't think you need polymorphism between the usages and quota APIs since they are separate APIs that serve different needs. I thought from discussions you just wanted some level of polymorphism within each API to represent resource type specific properties

rahuls-microsoft replied:
I will remove the polymorphism between usages and quota API.
Leave it for quota/limit, so that we can add addition properties, when MG quota is onboarded.

rahuls-microsoft replied:
Separated quotas and usages from  the polymorphism .

rahuls-microsoft committed:
Updating examples and restoring the api name to quotaLimitsProviders.

rahuls-microsoft committed:
Updating usages model and examples.

rahuls-microsoft committed:
Update quotaLimitProviders to remove resourceQuery.

rahuls-microsoft committed:
Revert "Update quotaLimitProviders to remove resourceQuery."

This reverts commit 301092d496b2834d2e5110aa7b7f32c6f9802951.

rahuls-microsoft committed:
Updating Operation API example for quotaLimitProviders rename.

rahuls-microsoft committed:
Removing quotaLimitProviders API, as it's not needed by the azure portal.
There wasn't a use case for the end user, that's why our team decided to remove it from public API.

rahuls-microsoft committed:
1. Removing unused models for quotaLimitProviders.
2. Update operations example.

rahuls-microsoft committed:
Remove allof from usages.

rahuls-microsoft commented:
https://msazure.visualstudio.com/One/_workitems/edit/10588364 - The breaking change has been approved the ARM breaking change review team.

rahuls-microsoft committed:
1. Updating 202 response schema.
2. Adding location header as per - https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#updating-using-patch

rahuls-microsoft committed:
Updating Patch 202 response model type.

rahuls-microsoft committed:
Updating description of put and API api to check location header for detailed status.

rahuls-microsoft committed:
Updates to description.

pilara-ms committed:
Update quota.json

pilara-ms committed:
Update quota.json

Text scrub up to line 256

pilara-ms committed:
Update quota.json

Scrubbed all text strings.

rahuls-microsoft committed:
Running prettier.

rahuls-microsoft committed:
Updating Network examples to include resourceType.

rahuls-microsoft committed:
Updating Networking Examples.

rahuls-microsoft committed:
Remove 202 response body for Put and Patch.

rahuls-microsoft committed:
preRemove example - putMachineLearningServicesQuotaRequestDedicated.json

rahuls-microsoft committed:
Remove example - putMachineLearningServicesQuotaRequestDedicated.json

rahuls-microsoft committed:
Merge branch 'master' of https://github.com/rahuls-microsoft/azure-rest-api-specs

BigCat20196 commented:
/azp run
